# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: layout/generic/nsIFrame.cpp
# Commit: ae6b08b3f4de
# Full Hash: ae6b08b3f4def1cf7cf12b8cfa53d0edaad99f97
# Author: Miko Mynttinen <mikokm@gmail.com>
# Date: 2022-02-23 09:38:00
# Regressor Bug: 1714584
# File Overlap Count: 1
# Description:
#   Bug 1714584 - Part 2: Remove nsDisplayList::RemoveBottom() r=mstange
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D138153
# ==============================================================================

diff -r a261bb633790 -r ae6b08b3f4de layout/generic/nsIFrame.cpp
--- a/layout/generic/nsIFrame.cpp	Tue Feb 22 23:42:18 2022 +0000
+++ b/layout/generic/nsIFrame.cpp	Tue Feb 22 23:42:18 2022 +0000
@@ -3702,7 +3702,8 @@
 
     nsDisplayItem* separator = nullptr;
 
-    while (nsDisplayItem* item = resultList.RemoveBottom()) {
+    // TODO: This can be simplified: |participants| is just |resultList|.
+    for (nsDisplayItem* item : resultList.TakeItems()) {
       if (ItemParticipatesIn3DContext(this, item) &&
           !item->GetClip().HasClip()) {
         // The frame of this item participates the same 3D context.
@@ -3883,7 +3884,8 @@
       container = MakeDisplayItem<nsDisplayContainer>(
           aBuilder, this, containerItemASR, &resultList);
     } else {
-      container = resultList.RemoveBottom();
+      MOZ_ASSERT(resultList.Length() == 1);
+      resultList.Clear();
     }
 
     // Mark the outermost display item as reusable. These display items and
@@ -3923,7 +3925,8 @@
   // invalidation) or we're doing a full build and don't need a wrap list, then
   // we can skip adding one.
   if (aBuiltContainerItem || (!aBuilder->IsPartialUpdate() && !needsWrapList)) {
-    aList->RemoveBottom();
+    MOZ_ASSERT(aList->Length() == 1);
+    aList->Clear();
     return item;
   }
 
@@ -3940,7 +3943,8 @@
     if (needsWrapList) {
       DiscardOldItems(aFrame);
     } else {
-      aList->RemoveBottom();
+      MOZ_ASSERT(aList->Length() == 1);
+      aList->Clear();
       return item;
     }
   }